アカウント機能の実装

tak848 opened the pull request:
django-allauthを用いて，アカウント機能を実装しました。具体的なメイン機能は，
- 新規登録
- ログイン
- ログアウト
の3点です。さらにメールアドレスでの認証機能も付いています。

プロフィールの変更も可能になっています。

最低限のローカル情報を別ファイルに退避させているため，再現時は[こちら](https://github.com/TAK848/pg-backend-final-tatter/blob/main/README.md)を確認いただければと思います。

本来もう少し機能ごとに分割してリクエストするべきだったと思います。
また，formの出力を変えるため，特にテンプレートがわかりにくくなってしまっています。すみません。
読みづらいプログラムで申し訳ございませんが，よろしくお願い致します。

/signup/
<img width="1616" alt="Screen Shot 2021-08-22 at 2 38 22" src="https://user-images.githubusercontent.com/41906969/130330603-37b6cd5d-b392-442b-99ce-2753f1cc1fa7.png">
/login/
<img width="1619" alt="Screen Shot 2021-08-22 at 2 38 09" src="https://user-images.githubusercontent.com/41906969/130330595-217b5fff-a757-4304-8e01-a7c5e8f2bc6f.png">
/
<img width="1617" alt="Screen Shot 2021-08-22 at 2 47 19" src="https://user-images.githubusercontent.com/41906969/130330686-38db0b59-9960-4b01-ab65-3bb924b83e76.png">
/user/TAK848/ （小文字でも可，他ユーザーもアクセス可）
<img width="1619" alt="Screen Shot 2021-08-22 at 2 49 02" src="https://user-images.githubusercontent.com/41906969/130330726-8944062b-78c9-4657-9fd3-7cfe8d6073ab.png">
/mypage/edit/
<img width="1619" alt="Screen Shot 2021-08-22 at 2 48 45" src="https://user-images.githubusercontent.com/41906969/130330721-26a50eb3-c1bf-4174-a48a-4fd1e3de2689.png">

tak848 committed:
settings.pyの基本設定を追加・更新

tak848 committed:
[add]アカウント関係の機能を追加

django-allauthを利用し，新規登録・ログイン・ログアウト機能を実装。さらにメールアドレスの確認機能・パスワード変更機能を実装。ユーザーページを実装し，プロフィールを変更できるようにしました。

tak848 committed:
Merge branch 'dev-accounts' into accounts

tak848 committed:
settings.pyから使用していないアプリ（django-bootstrap5）を削除

TonkyH commented on accounts/adapter.py:
```diff
@@ -0,0 +1,9 @@
+        user = super().save_user(request, user, form, commit=False)
+        user.handle_name = form.cleaned_data.get('handle_name')
+        user.set_random_username()
+        user.save()
```
```python
return user
```
こんな感じで
最後`user`を返してないですが問題ないですか？？

tak848 replied:
継承元にならって足しておきました！

TonkyH commented on accounts/adapter.py:
```diff
@@ -0,0 +1,9 @@
+
+class CustomAccountAdapter(DefaultAccountAdapter):
+    def save_user(self, request, user, form, commit=True):
+        user = super().save_user(request, user, form, commit=False)
```
`commit=False`で平気ですか？？
`commit=commit`で受け取った値を使った方がいいように見えます！

tak848 replied:
allauthを使用しておきながらAdapterの仕様をあまり理解していないので，[Qiitaのこのページ](https://qiita.com/Kei-Segawa/items/aa38f72fc5409a3517c6#8adapterpy)を参考に書きました。
試しに`commit=commit`とすると`True`が入ってきて，そのまま保存しようとされるみたいです。しかし，`display_username`が未設定で，adaper内でランダム設定しているため，この段階では何もなく，IntegrityErrorでUNIQUEじゃないよ！とエラーが出てしまいますので`False`として最後にcommitするように（解釈した結果的には）なっています。

TonkyH replied:
参考資料ありがとうございます！

継承元 の`save_user()`を`commit=False`で呼び出して、`user`のインスタンスを作成するので
これでOKでした🙇‍♂️🙇‍♂️

TonkyH commented on accounts/forms.py:
```diff
@@ -0,0 +1,66 @@
+
+    def clean_handle_name(self):
+        value = self.cleaned_data.get('handle_name')
+        print('handle_name')
+        print(value)
```
不要なら消しましょう！！

tak848 replied:
すみません！削除しました！！

TonkyH commented on accounts/forms.py:
```diff
@@ -0,0 +1,66 @@
+    uuid = forms.CharField()
+
+    class Meta:
+        model = get_user_model()
```
`get_user_model()`は
```python
User = get_user_model()

class CustomSignupForm(SignupForm):
```
みたいに上で定義して、`User`として使いまわすやり方をよく見るので、そっちの方がいいかも？？

この辺は好みが出ると思うので、いいね！って感じでしたら修正しておいてください

```python
model = User
```
こんな感じになると思います！

tak848 replied:
修正しました！よくわからず突貫で書いていました。わかりやすく書きやすいです。
ありがとうございます！！

TonkyH replied:
他と同様、`User`にしましょう！

TonkyH commented on accounts/forms.py:
```diff
@@ -0,0 +1,66 @@
+        return self.cleaned_data.get('display_username').strip().lstrip('@').strip()
+
+    def clean(self):
+        users = get_user_model()
```
```python
User = get_user_model()
```
これを定義しておけば、
`users = get_user_model()`はいらなくなります

tak848 replied:
修正しました！！

TonkyH commented on accounts/forms.py:
```diff
@@ -0,0 +1,66 @@
+        users = get_user_model()
+        uuid = self.cleaned_data.get('uuid')
+        try:
+            user = users.objects.get(uuid=uuid)
```
そうすると、ここで、
```python
user = User.objects.get(uuid=uuid)
```
みたいにできますので、他のモデルと書き方を揃えることができます！

TonkyH replied:
こんな感じのがちょいちょいあるので、確認してみてください！

tak848 replied:
修正しました！
その他も検索して修正しました！！

TonkyH commented on accounts/views.py:
```diff
@@ -1,3 +1,107 @@
+    #     return super().form_valid(form)
+
+    # def get_form_kwargs(self):
+    #     kwargs = super().get_form_kwargs()
+    #     # 更新前のユーザー情報をkwargsとして渡す
+    #     # kwargs.update({
+    #     #     'biography': self.request.user.biography,
+    #     # })
+    #     return kwargs
+    # def get(self, request):
+    #     context = {}
+    #     return TemplateResponse(request, 'account/mypage.html', context)
+
+    # def post(self, request):
+    #     context = {}
+    #     # if request.user.check_password(request.POST.get('password')):
+    #     if True:
+    #         # user = get_user_model().objects
+    #         user = request.user
+    #         user.biography = 'aasaa\n\n\n\n\n\na'
+    #         if user.is_valid():
+    #             user.save()
+    #         # request.user.save()
+    #         # form = ProfileForm(request.POST)
+    #         # if form.is_valid():
+    #         #     print('valid')
+    #         #     user = form.save(commit=False)
+    #         #     print('ok')
+    #         #     user.save()
+    #         #     print('save')
+    #         #     messages.success(request, 'プロフィールを更新しました')
+    #         #     # request.user.biography = 'unko'
+    #         # else:
+
+    #         #     messages.error(request, 'プロフィール更新失敗?')
+    #     else:
+    #         context['password_check'] = False
+    #         messages.error(request, 'プロフィール更新失敗')
+
+    #     return TemplateResponse(request, 'account/mypage.html', context)
```
不要なら消してください

tak848 replied:
すみません！削除しました！！

TonkyH commented on accounts/views.py:
```diff
@@ -1,3 +1,107 @@
+
+
+class CustomSignupView(SignupView):
+    def dispatch(self, request, *args, **kwargs):
+        if (request.method == 'POST' and
+                '/signup/' in request.get_full_path(True)):
+            self.was_post = True
+        else:
+            self.was_post = False
+        return super().dispatch(request, *args, **kwargs)
+
```
ここでは何をしてますか？？

TonkyH commented on accounts/views.py:
```diff
@@ -1,3 +1,107 @@
+    template_name = 'account/mypage_edit.html'
+    model = get_user_model()
+    form_class = UserForm
+    # fields = ('biography',)
```
不要なら消しましょう

tak848 replied:
すみません！削除しました！！

TonkyH commented on accounts/views.py:
```diff
@@ -1,3 +1,107 @@
+
+class MypageView(LoginRequiredMixin, UpdateView):
+    template_name = 'account/mypage_edit.html'
+    model = get_user_model()
```
別の場所でコメントしたよに
`User = get_user_model()`を定義して`User`を使いまわした方がわかりやすいかも

TonkyH commented on accounts/views.py:
```diff
@@ -1,3 +1,107 @@
+    # fields = ('biography',)
+    success_url = reverse_lazy('accounts:mypage_edit')
+
+    def get_object(self):
+        return self.request.user
```
これは必要ですか？？

tak848 replied:
再度調べてみたら，テンプレート内から`request`を参照できるとわかったので，`{% if request.method != 'GET' ... %}`と書き換えることで削除しました！

目的は，サーバーへのPOST後のエラー返却時にのみ，form内のinputのclassに`is-valid`や`is-invalid`を入れることでした。書いていた時は方法がわからずこうしていました。（テンプレート内で`{% if was_post ... %}`のように利用）
`'/signup/' in request.get_full_path(True)`のようにパスをチェックしていたのは，全く別の場所からPOSTかけられた時の対策のつもりでした。ですが対策しなくても普通のユーザーは困りませんし何よりCSRFも付いててセキュリティ上もほぼ問題ないので同時に削除しました。

TonkyH replied:
しっかり調べてくれてありがとうございます！！！
勉強になります！

TonkyH commented on tatter/urls.py:
```diff
@@ -0,0 +1,10 @@
+app_name = 'tatter'
+urlpatterns = [
+    path('', views.IndexView.as_view(), name='index'),
+    re_path(r'^user/(?P<username>\w+)/$',
+            views.UserProfileView.as_view(), name='user_profile'),
```
```python
path('user<str:username>/', views.UserProfileView.as_view(), name='user_profile')
```
これだとダメでしたか？？

tak848 replied:
すみません，ユーザー名に英数字に加えアンダーバー（_）も許可しようと思い正規表現を使おうとしていましたが，書き忘れていましたので，`\w`から`[\w_]`に修正しました！

TonkyH commented on tatter/views.py:
```diff
@@ -1,3 +1,29 @@
+        context = {
+            'search_username': username,
+        }
+        try:
+            user = users.objects.get(username=username.lower())
+            context['exist_user'] = True
+            if user == request.user:
+                context['is_me'] = True
+            context['got_user'] = user
+        except users.DoesNotExist:
+            context['exist_user'] = False
+        return TemplateResponse(request, 'tatter/user_profile.html', context)
```
処理の流れ的にこっちの方が追いやすいような気がします
そんなことなかったらごめんなさい🙇‍♂️
```python
try:
    user = users.objects.get(username=username.lower())
    context['exist_user'] = True
except users.DoesNotExist:
    context['exist_user'] = False
    return TemplateResponse(request, 'tatter/user_profile.html', context)    

if user == request.user:
    context['is_me'] = True
context['got_user'] = user
return TemplateResponse(request, 'tatter/user_profile.html', context)
```

あと、`is_me`, `got_user`が入ったり入らなかったりするのは問題ないですか？？

tak848 replied:
try...exceptは短い方がわかりやすいので修正しました！！
（Pythonのスコープはtry...except内に書いても外に出てくるのでuserも一応大丈夫ですし）

テンプレート内の処理的には問題ないです。
ifの結果がNoneならFalse判定されますが，一部の他言語で同じことをすると怒られそうなのでcontextの宣言内にデフォルト値を設定しておきました。

TonkyH reviewed (commented):
確認お願いします！

tak848 committed:
一部テンプレートファイルのタブ幅の修正，その他軽微な修正

一部テンプレートのタブ幅が4になっていたので2に修正しました。viewからcontextとして受け取っていたwas_postを，request.method=='POST'に変更しました。その他，軽微な修正をしました。

tak848 committed:
アカウント機能にて，レビューいただいた修正点を修正しました。

tak848 committed:
テンプレートにて，自己紹介の文字制限を明記

tak848 committed:
ユーザーページのviewのcontextのデフォルト値を設定

TonkyH commented on accounts/models.py:
```diff
@@ -1,3 +1,150 @@
+        return f'{self.handle_name}@{self.username}<{self.email}>'
+
+    def set_random_username(self, n=12):
+        user = get_user_model()
```
他と同様、`User`にしましょう！

TonkyH commented on accounts/models.py:
```diff
@@ -1,3 +1,150 @@
+
+    def clean(self):
+        super().clean()
+        user = get_user_model()
```
他と同様、`User`にしましょう！

tak848 committed:
User取得を統合

fumiharun11 commented on accounts/models.py:
```diff
@@ -1,3 +1,150 @@
+
+from .validate import linebreak_limit_biography
+
+User = get_user_model()
```
初回`./manage.py migrate`実行時にここが原因でエラーが発生します。
原因は容易に特定できると思うので、ローカルDBを削除するか一旦新規にリポジトリをcloneするなどして検証してみてください。
```
Traceback (most recent call last):
...
    return self.models[model_name.lower()]
KeyError: 'customuser'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
...
    raise LookupError(
LookupError: App 'accounts' doesn't have a 'CustomUser' model.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
...
    raise ImproperlyConfigured(
django.core.exceptions.ImproperlyConfigured: AUTH_USER_MODEL refers to model 'accounts.CustomUser' that has not been installed
```

tak848 replied:
まだmigrateしてCustomUserを作成していない段階で，グローバルで`get_user_model()`を呼び出しているためエラーを吐いているというのがすぐ特定できました。
ひとまず使うところで逐一呼び出す方式に戻したら，migrateがうまく行きました。
（もしもっと良い方法がありましたら教えていただけると幸いです。）

tak848 committed:
[fix]初回にmigrateできない問題を修正

tak848 committed:
Merge branch 'dev-accounts' into accounts

fumiharun11 commented on accounts/forms.py:
```diff
@@ -0,0 +1,65 @@
+        return super().clean_login().lstrip('@').strip()
+
+
+class UserForm(forms.ModelForm):
```
このフォームはユーザー情報更新用のフォームですか？
用途が限られていればclass名に含めると視認性が上がると思います！

tak848 replied:
今のところその通りなので修正しました！

fumiharun11 commented on accounts/models.py:
```diff
@@ -1,3 +1,150 @@
+
+    def create_user(self, email, password=None):
+        if not email:
+            raise ValueError('Users must have an email address')
```
ここのValueErrorは英文渡しているけど想定通り？

tak848 replied:
変更を忘れてたので修正しました！

fumiharun11 commented on accounts/models.py:
```diff
@@ -1,3 +1,150 @@
+        while True:
+            made_username = ''.join([random.choice(dat) for i in range(n)])
+            try:
+                if User.objects.get(username=made_username.lower()):
```
例外処理でエラーをキャッチした場合に処理を実行する方針にした理由はありますか？
以下のように存在判定クエリををしてif分岐させるだけでも十分な印象があります。
```py
User.objects.filter(username=made_username.lower()).exists()
```
参考: https://docs.djangoproject.com/en/3.2/ref/models/querysets/#django.db.models.query.QuerySet.exists

tak848 replied:
とりあえず動け！って勢いで書いてました…
補足でいただいた提案から作りました，`make_random_username`に移動し修正しました！
公式リファレンスをもっと読みたいと思います！

fumiharun11 commented on accounts/models.py:
```diff
@@ -1,3 +1,150 @@
+            except User.DoesNotExist:
+                self.username = made_username.lower()
+                self.display_username = made_username
+                return made_username
```
returnする必要のあるメソッド？

tak848 replied:
どこでも使っていませんでした…
最終的には補足でいただいたように，`make_random_username`関数を作り，`set_new_random_username`へreturnするようにしました！

fumiharun11 commented on accounts/models.py:
```diff
@@ -1,3 +1,150 @@
+        return f'{self.handle_name}@{self.username}<{self.email}>'
+
+    def set_random_username(self, n=12):
+        User = get_user_model()
```
このままの方針でも良いですが、

1. CustomUserをそのまま使用する
2. `self.__class__`からクラスを取得する

などがあるので紹介しておきます。

fumiharun11 replied:
### 補足
このメソッドの複雑なロジックは、ここに持たせず別の関数(`make_random_username`)や別のクラスのメソッドに分解し、ここではsetterとしての責務のみにすると可読性が上がると思います。

tak848 replied:
自分の感覚的には`self.__class__`が一番しっくりきたのでこれを採用させていただきました！

補足について:
`set_new_random_username`と`make_random_username`に分けることで対応してみました！

fumiharun11 commented on accounts/models.py:
```diff
@@ -1,3 +1,150 @@
+    def clean(self):
+        super().clean()
+        User = get_user_model()
+        if not self.username == self.display_username.lower():
```
細かいけど`self.display_username.lower()`を何回も呼び出すのであれば変数に保持しておくと良いと思います
```py
lower_display_username = self.display_username.lower()
```

tak848 replied:
修正しました！！！

fumiharun11 commented on config/context_processors.py:
```diff
@@ -0,0 +1,5 @@
+def my_context_processor(req):
+    return {
+        'domain_name': 'https :// test tatter . com（仮）',
```
いずれ変更予定だと思うけど、このフォーマットだとドメイン以外にプロトコルも含んでいるのでkey名を変更するかvalueを適切なフォーマットにすると混乱しなくて:+1:

tak848 replied:
allauthのメールのテンプレートに使われていたのを何も考えずに上書きしてました。修正しました！

fumiharun11 commented on config/settings.py:
```diff
@@ -55,14 +67,17 @@
         'BACKEND': 'django.template.backends.django.DjangoTemplates',
-        'DIRS': [],
+        'DIRS': [
+            os.path.join(BASE_DIR, 'templates'),
```
BASE_DIRの型が`pathlib.PosixPath`なので`os.path`を使用しない方針で良いと思ったのだけどどうかな？
```py
BASE_DIR / 'templates'
```

tak848 replied:
とりあえず何処かからコピペしてました。良いと思うので修正しました！！

fumiharun11 commented on config/settings.py:
```diff
@@ -119,16 +140,61 @@
 
 STATIC_URL = '/static/'
+STATICFILES_DIRS = [
+    os.path.join(BASE_DIR, 'static'),
```
TEMPLATESと同様のコメント

tak848 replied:
修正しました！

fumiharun11 commented on tatter/urls.py:
```diff
@@ -0,0 +1,10 @@
+app_name = 'tatter'
+urlpatterns = [
+    path('', views.IndexView.as_view(), name='index'),
+    re_path(r'^user/(?P<username>[\w_]+)/$',
```
ここで`re_path`でpath生成する方針にしたのは何故ですか？

tak848 replied:
Twitterのユーザー名が半角英数もしくはアンダーバーのみで，それをパクろうとしたのですが，slugを使うとハイフンが含まれてしまうので，仕方なく正規表現を利用するようにしました…

fumiharun11 replied:
`re_path`は要件に合わせて使用しているのですね。オッケーです:+1:
正規表現に対してコメントですが、`[\w_]`で`_`を含める必要はありましたか？

tak848 replied:
ありませんでした…！
修正しました！

fumiharun11 commented on accounts/forms.py:
```diff
@@ -0,0 +1,65 @@
+    )
+    agreement = forms.BooleanField(label='利用規約に同意')
+
+    def signup(self, request, user):
```
このメソッドは`SignupForm`に定義(又は継承)されているメソッドのオーバーライド？

tak848 replied:
記憶の限りは[ここ](https://qiita.com/Kei-Segawa/items/aa38f72fc5409a3517c6)をコピったのだと思います…継承元を見ても存在してませんでしたし，意味もないので削除しました！！

fumiharun11 reviewed (commented):
全体の構成として基本的に問題なく良いと思います！
Slackで共有済みではありますが、今後PR出す時は機能単位で出すようにし、差分が大きくなる時は事前に相談をお願いします〜！

tak848 committed:
いただいた修正点に加え，settingsのLOGIN・LOGOUT関係のurlの参照先を変更

tak848 committed:
正規表現を修正

fumiharun11 reviewed (approved):
![](https://media.tumblr.com/b999d6fa94d40545edb2dac70623e070/tumblr_inline_n2ztfg6zr01qaxsyz.gif)
